Lib/logging/handlers.py

```diff
@@ -78,7 +79,13 @@ def emit(self, record):
         try:
             if self.shouldRollover(record):
                 self.doRollover()
-            logging.FileHandler.emit(self, record)
+            if self._msg:
+                msg = self._msg
+                self._msg = None
+            else:
+                msg = self.format(record)
+            self.stream.write(msg + self.terminator)
```

Comment:
`FileHandler.emit()` opens a stream if it was not open. If you want to use the cached formatted message, this should perhaps be done in `FileHandler.emit()` or `StreamHandler.emit()`. Or inline `FileHandler.emit()` and `StreamHandler.emit()` completely here. Or add parameter for formatted error message in `FileHandler.emit()` and `StreamHandler.emit()`.

Reply:
> FileHandler.emit() opens a stream if it was not open

True, but it's opened in `shouldRollover()`, so that part of `FileHandler.emit()` would never be exercised.
https://github.com/python/cpython/blob/9e551f9b351440ebae79e07a02d0e4a1b61d139e/Lib/logging/handlers.py#L196-L197

Reply:
You are right. Although it may break subclasses that override `shouldRollover()` (if it returns early, before calling `super().shouldRollover()`).